src/util/text-tracks.js

```diff
@@ -223,6 +223,60 @@ export const addMetadata = ({
   });
 };
 
+/**
+ * Add daterange metadata text track to a source handler given an array of metadata
+ *
+ * @param {Object}
+ *   @param {Object} inbandTextTracks the inband text tracks
+ *   @param {Array} metadataArray an array of meta data
+ * @private
+ */
+export const addDaterangeMetadata = ({
+  inbandTextTracks,
+  metadataArray,
+  timestampOffset
+}) => {
+  if (!metadataArray) {
+    return;
+  }
+
+  const Cue = window.WebKitDataCue || window.VTTCue;
+  const metadataTrack = inbandTextTracks.metadataTrack_;
+  const daterangeAttr = {
+    id: 'ID',
+    class: 'CLASS',
+    startDate: 'START-DATE',
+    duration: 'DURATION',
+    endDate: 'END-DATE',
+    endOnNext: 'END-ON-NEXT',
+    plannedDuration: 'PLANNED-DURATION',
+    scte35Out: 'SCTE35-OUT',
+    scte35In: 'SCTE35-IN'
+  };
+
+  if (!metadataTrack) {
+    return;
+  }
+
+  metadataArray.daterange.forEach((metadata) => {
+    const startTime = timestampOffset + (new Date(metadata.startDate).getTime() - new Date(metadataArray.dateTimeObject).getTime()) / 1000;
+    const endTime = metadata.endDate ? (new Date(metadata.endDate).getTime() - new Date(metadataArray.dateTimeObject).getTime()) / 1000 : 0;
```

Comment:
IMHO, here are a few naming ideas:
- Instead of`metadataArray`, the better name is `mediaPlaylist`, because it is parsed media playlist and not an array of metadata.
- Instead of `daterange`, the better name is `dateRanges` because it is literally `Array<DateRange>` and it is semantically right to say `dateRanges.map((dateRange) => ....)`

Regarding `endTime` I would assume something like this (we can use backward loop to optimize performance, but I keep 2 for loops for simplicity):

```javascript
    const localDateRanges = [...mediaPlaylist.dateRanges];
    const dateRangeClasses = {};

    for (let dateRange of localDateRanges) {
        const startDateTimeStamp = new Date(metadata.startDate).getTime();
        const startTime = timestampOffset + (startDateTimeStamp - dateTimeTimeStamp) / 1000;

        dateRange.startTime = startTime;

        if (dateRange.class) {
            if (dateRangeClasses[dateRange.class]) {
                dateRangeClasses[dateRange.class].push(dateRange);
            } else {
                dateRangeClasses[dateRange.class] = [dateRange];
            }
        }
    }

    for (let dateRange of localDateRanges) {
        let endTime;
        const classList = dateRangeClasses[dateRange.class];
        const classListIndex = classList.findIndex(dateRange);

        if (dateRange.endDate) {
            // Use endDate directly
            const endDateTimeStamp = new Date(metadata.startDate).getTime();
            endTime = timestampOffset + (endDateTimeStamp - dateTimeTimeStamp) / 1000;
        } else if (dateRange.endOnNext && classList[classListIndex + 1]) {
            // Fallback to the startTime of the next dateRange with the same class
            const { startTime } = classList[classListIndex + 1];
            endTime = startTime;
        } else if (dateRange.duration) {
            // Fallback to startTime + duration
            endTime = dateRange.startTime + dateRange.duration;
        } else if (dateRange.plannedDuration) {
            // Fallback to startTime + plannedDuration
            endTime = dateRange.startTime + dateRange.plannedDuration;
        } else {
            // Fallback to 0
            endTime = 0;
        }

        // generate and add cue
    }
```

Reply:
Thank you! Did the renaming and updated endTime in the latest commit.